src/main/java/com/codecool/shop/controller/ProductController.java

```diff
@@ -2,39 +2,211 @@
 
 import com.codecool.shop.dao.ProductCategoryDao;
 import com.codecool.shop.dao.ProductDao;
+import com.codecool.shop.dao.SupplierDao;
 import com.codecool.shop.dao.implementation.ProductCategoryDaoMem;
 import com.codecool.shop.dao.implementation.ProductDaoMem;
 import com.codecool.shop.config.TemplateEngineUtil;
+import com.codecool.shop.model.Cart;
+import com.codecool.shop.dao.implementation.SupplierDaoMem;
+import com.codecool.shop.model.Product;
+import com.codecool.shop.model.ProductCategory;
+import com.codecool.shop.model.Supplier;
 import org.thymeleaf.TemplateEngine;
 import org.thymeleaf.context.WebContext;
-
 import javax.servlet.ServletException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.io.IOException;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.ArrayList;
+import java.util.List;
 
 @WebServlet(urlPatterns = {"/"})
 public class ProductController extends HttpServlet {
 
+    private StringBuffer categoryNames = new StringBuffer();
+    private StringBuffer supplierNames = new StringBuffer();
+    private boolean requestForCategory = false;
+    private boolean requestForSupplier = false;
+
     @Override
     protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+
+        categoryNames.delete(0, categoryNames.length());
+        supplierNames.delete(0, supplierNames.length());
+        requestForSupplier = false;
+        requestForCategory = false;
+
         ProductDao productDataStore = ProductDaoMem.getInstance();
         ProductCategoryDao productCategoryDataStore = ProductCategoryDaoMem.getInstance();
+        SupplierDao supplierDataStore = SupplierDaoMem.getInstance();
+
+        List<Product> products = filterProduct(req, supplierDataStore, productCategoryDataStore, productDataStore);
```

Comment:
Do not pass the whole request around! Only those info which is needed to execute the operation.

Reply:
Store daos in a fields so you don't need to pass them around